tests/Unit/Helpers/DataStructures/Tensor/Expressions/EvaluateRank2TestHelpers.hpp

```diff
@@ -0,0 +1,153 @@
+///
+/// We test nonsymmetric indices and symmetric indices across two functions to
+/// ensure that the code works correctly with symmetries. This function tests
+/// one of the following symmetries:
+/// - <2, 1> (`test_evaluate_rank_2_no_symmetry`)
+/// - <1, 1> (`test_evaluate_rank_2_symmetric`)
+///
+/// \details `TensorIndexA` and `TensorIndexB` can be any type of TensorIndex
+/// and are not necessarily `ti_a_t` and `ti_b_t`. The "A" and "B" suffixes just
+/// denote the ordering of the generic indices of the RHS tensor expression. In
+/// the RHS tensor expression, it means `TensorIndexA` is the first index used
+/// and `TensorIndexB` is the second index used.
+///
+/// \tparam DataType the type of data being stored in the Tensors
+/// \tparam TensorIndexTypeA the \ref SpacetimeIndex "TensorIndexType" of the
+/// first index of the RHS Tensor
+/// \tparam TensorIndexTypeB the \ref SpacetimeIndex "TensorIndexType" of the
+/// second index of the RHS Tensor
+/// \tparam ValenceA the valence of the first index used on the RHS of the
+/// TensorExpression
+/// \tparam ValenceB the valence of the second index used on the RHS of the
+/// TensorExpression
+/// \param tensorindex_a the first TensorIndex used on the RHS of the
+/// TensorExpression, e.g. `ti_a`
+/// \param tensorindex_b the second TensorIndex used on the RHS of the
+/// TensorExpression, e.g. `ti_B`
+template <
+    typename DataType, template <size_t, UpLo, typename> class TensorIndexTypeA,
+    template <size_t, UpLo, typename> class TensorIndexTypeB, UpLo ValenceA,
+    UpLo ValenceB, typename TensorIndexA, typename TensorIndexB>
+void test_evaluate_rank_2_no_symmetry(
+    const TensorIndexA& tensorindex_a,
+    const TensorIndexB& tensorindex_b) noexcept {
+#define DIM_A(data) BOOST_PP_TUPLE_ELEM(0, data)
+#define DIM_B(data) BOOST_PP_TUPLE_ELEM(1, data)
+#define FRAME(data) BOOST_PP_TUPLE_ELEM(2, data)
+
+#define CALL_TEST_EVALUATE_RANK_2_CORE(_, data)                          \
+  test_evaluate_rank_2_core<                                             \
+      DataType, Symmetry<2, 1>,                                          \
+      index_list<TensorIndexTypeA<DIM_A(data), ValenceA, FRAME(data)>,   \
+                 TensorIndexTypeB<DIM_B(data), ValenceB, FRAME(data)>>>( \
+      tensorindex_a, tensorindex_b);
+
+  GENERATE_INSTANTIATIONS(CALL_TEST_EVALUATE_RANK_2_CORE, (1, 2, 3),
+                          (1, 2, 3), (Frame::Grid, Frame::Inertial))
+
+#undef CALL_TEST_EVALUATE_RANK_2_CORE
+#undef FRAME
+#undef DIM_B
+#undef DIM_A
+}
+
+/// \copydoc test_evaluate_rank_2_no_symmetry()
+template <
+    typename DataType, template <size_t, UpLo, typename> class TensorIndexTypeA,
+    template <size_t, UpLo, typename> class TensorIndexTypeB, UpLo ValenceA,
+    UpLo ValenceB, typename TensorIndexA, typename TensorIndexB>
+void test_evaluate_rank_2_symmetric(
+    const TensorIndexA& tensorindex_a,
+    const TensorIndexB& tensorindex_b) noexcept {
```

Comment:
How do you feel about me instead changing the test to only take one template parameter for the (shared) `TensorIndexType` (which is `SpatialIndex` or `SpacetimeIndex`) and one (shared) valence? Meaning, change
```
template <
    typename DataType, template <size_t, UpLo, typename> class TensorIndexTypeA,
    template <size_t, UpLo, typename> class TensorIndexTypeB, UpLo ValenceA,
    UpLo ValenceB, typename TensorIndexA, typename TensorIndexB>
void test_evaluate_rank_2_symmetric(
    const TensorIndexA& tensorindex_a,
    const TensorIndexB& tensorindex_b)
```
to
```
template <
    typename DataType, template <size_t, UpLo, typename> class TensorIndexType,
    UpLo Valence, typename TensorIndexA, typename TensorIndexB>
void test_evaluate_rank_2_symmetric(
    const TensorIndexA& tensorindex_a,
    const TensorIndexB& tensorindex_b)
```

As far as ensuring that `test_evaluate_rank_2_impl` is given the same (1) `TensorIndexType`s (`SpatialIndex` or `SpacetimeIndex` ) that also have the same (2) dimension, (3) valence, and (4) frame type, I  think (1) and (3) would be solved by the above, and (2) and (4) are already addressed since the `index_list` created uses the same dimension and frame type for both indices since they both use `DIM(data)` and `FRAME(data)` (see [here](https://github.com/sxs-collaboration/spectre/blob/b8a87e8ebcbd357ec61a553cd8defaed5cdebfac/tests/Unit/Helpers/DataStructures/Tensor/Expressions/EvaluateRank2TestHelpers.hpp#L141))

Reply:
I guess the only complication with this approach is that I wouldn't be able to use `\copydoc` to copy around the documentation because the `\tparam`s would be different (this has more of an impact on the rank 3 test functions since there are multiple). So I guess the decision is, which is better:
1) Reduce the template parameters as shown above, but now can only copy around the `\brief` and `\details` (instead of being able to wholesale copy documentation with `\copydoc` since the `tparam`s will be different)
2) Just put in static asserts, keeping redundant template parameters that are checked for equality

Let me know if you have a preference!

Reply:
- Ah, my comment was actually intended to question whether the two arguments to the function are of the same type, but it escaped me that -- as you've suggested for other template parameters -- you can ensure that `tensorindex_a` and `tensorindex_b` have the same type by replacing the two template parameters `TensorIndexA` and `TensorIndexB` with a single (shared) `TensorIndex`. This would definitely be cleaner than using a bunch of `static_asserts`!
- What you've suggested, of _also_ reducing the other template parameters to further enforce the symmetry, sounds like a good change as well! In terms of the choice you're suggesting, I'd personally opt for cleaner code at the expense of more documentation, i.e., option 1, but both would certainly be just fine. Do what you like best!

Reply:
Thanks! I attempted doing the first option, but ran into an issue that conflicts with a [previous change request](https://github.com/sxs-collaboration/spectre/pull/2475#discussion_r488906630) from @nilsdeppe . When I do the swap I mentioned in the [above comment](https://github.com/sxs-collaboration/spectre/pull/2475#discussion_r499100807), I updated the documentation of `test_evaluate_rank_2_symmetric` from simply:
```
/// \copydoc test_evaluate_rank_2_no_symmetry()
```

to:
```
/// \ingroup TestingFrameworkGroup
/// \copybrief test_evaluate_rank_2_no_symmetry()
///
/// \copydetails test_evaluate_rank_2_no_symmetry()
///
/// \tparam DataType the type of data being stored in the Tensors
/// \tparam TensorIndexType the shared \ref SpacetimeIndex "TensorIndexType" of
/// the indices
/// \tparam Valence the shared valence of the indices
/// \param tensorindex_a the first TensorIndex used on the RHS of the
/// TensorExpression, e.g. `ti_a`
/// \param tensorindex_b the second TensorIndex used on the RHS of the
/// TensorExpression, e.g. `ti_B`
```

This change was made because the template parameters of this symmetric function are now no longer the same as those for `test_evaluate_rank_2_no_symmetry`. However, the problem I'm seeing when viewing the documentation that this outputs is that `\copydetails` copies not only the `\details` section over, but also the `\tparam`s and `\param`s (the documentation shows both `\tparam`s and `\param`s from both functions). So then, the naive solution would be to manually copy and paste over the `\details` like I had before, but then at that point, if I'm writing the `\tparam`s, `param`s, and `\details` manually, that just leaves the `\brief` being copied. And since the `\brief`s originally were the only thing different between the functions that tested different symmetries, I don't know how much is gained from copying it with `\copybrief` vs just manually writing the briefs, as I originally did, because everything else in the functions' documentation will have to be manually written anyway. All of this is a little more imporrtant for the rank 3 functions, because there are 5 of those that call the `impl` function, but only 2 for the rank 2 case - meaning, more documentation to copy around vs manually copy + tweak.

Let me know if you two have thoughts on this. I'm happy to do whatever! I'm just not sure how to reconcile this because I think that removing the duplicate template parameters is a smart update, but I don't know how to update the documentation accordingly without conflicting with Nils' previous change request of not manually copying and pasting redundant documentation.

Reply:
(note: the above refers to the behavior I'm seeing locally - I haven't yet pushed a commit with these changes)

Reply:
I see the problem... I've run into a similar issue once and I ended up simply not documenting the template parameters. I'm not proud, but it got around the problem 😛.

One idea might be to continue using "copydoc" per Nils' suggestion, keep the change we've been discussing of simplifying the template parameters, but simply accept that the copied docs aren't completely accurate for the symmetric case. By adding to the documentation a short "Note: the function for symmetric tensors has fewer template parameters", it should hopefully be easy to see how to map from the documentation of the general case to the code with fewer template parameters?

I think I like this option better than the alternatives (writing the docs out twice, or keeping the duplicate template parameters). What do you and @nilsdeppe think?

Reply:
Sounds fine to me. I haven't gone through the full discussion in detail, so this might not be relevant, but keep in mind that you don't need to document deduced template parameters since the argument from which they are deduced is usually documented.

Reply:
Thanks so much for the input and feedback! I went ahead and made this change locally. However, the behavior I'm seeing now is that `\copydoc` isn't copying over the documentation when I do something like

```
/// \copydoc test_evaluate_rank_2_no_symmetry()
template <
    typename DataType, template <size_t, UpLo, typename> class TensorIndexType,
    UpLo Valence, typename TensorIndexA, typename TensorIndexB>
void test_evaluate_rank_2_symmetric(
```

There is no documentation for `test_evaluate_rank_2_symmetric` generated at all, as far as I can tell. I'm thinking that maybe this is due to the fact that `test_evaluate_rank_2_no_symmetry` and `test_evaluate_rank_2_symmetric` now have different template parameters, so `\copydoc` isn't able to correctly find the function (link object) from which to copy documentation. I tried searching doxygen's website and GitHub issues to see if there was a way to specialize the copying with template parameters, but I haven't found anything. Do either of you know if this is possible? And if there isn't a way to use `\copydoc` anymore, do you have a preferred alternative option?

Reply:
I'm not familiar enough of with Doxygen to know the answers here. One alternative might be (I think the two functions are back-to-back in the code?) to try using Doxygen's "@" syntax to assign the documentation to the two functions, like this:
```
// @{
/// doxygen comment
template<...> void func1 {...}
template<...> void func1 {...}
// @}

But I'm not sure if that will work for this case with the different template parameters, because it might just be different syntax for the same underlying copy action, and might fail in the same cases.

Reply:
Hmm I might actually be doing something wrong on my end. I reverted my changes back to when all of the template parameters were the same and used `\copydoc` and I found the same problem. Something else is preventing `\copydoc` from working correctly and I'm going to look into it. Just wanted to made sure I shared this here before you or Nils went out of your way to think of a work-around.

Reply:
FWIW, I was able to fix this issue by prepending with `/// \ingroup TestingFrameworkGroup`, like so:
```
/// \ingroup TestingFrameworkGroup
/// \copydoc test_evaluate_rank_2_no_symmetry()
```